### ISSUE TITLE
Failed to run ASP.NET Core Runtime 6 application with Gramine

### ISSUE THREAD

**Issue by mssun** [gramineproject/gramine#1000]
### Description of the problem

I cannot run ASP.NET Core Runtime 6 "hello world" [sample application](https://github.com/Azure-Samples/dotnetcore-docs-hello-world) with Gramine.

### Create `Dockerfile` with the following content
```
$ cat Dockerfile
FROM gramineproject/gramine:v1.3 AS gramine

# Prepare the environment
ENV DEBIAN_FRONTEND=noninteractive
ENV TZ=America/Los_Angeles
RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
RUN apt-get update && apt-get install -y gcc make
RUN curl https://packages.microsoft.com/config/ubuntu/20.04/packages-microsoft-prod.deb -o packages-microsoft-prod.deb && \
    dpkg -i packages-microsoft-prod.deb && \
    rm packages-microsoft-prod.deb && \
    apt-get update && \
    apt-get install -y aspnetcore-runtime-6.0

# Build the dotnet app
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /source
COPY . ./
WORKDIR /source/dotnetcore-docs-hello-world
RUN dotnet publish -c release -o /app/publish

# Finnal running enviroment
FROM gramine AS final
WORKDIR /app
COPY --from=build /app/publish ./publish
COPY dotnet.manifest.template .
COPY Makefile .
RUN gramine-sgx-gen-private-key
RUN make clean && SGX=1 DEBUG=1 make
EXPOSE 80

ENTRYPOINT ["gramine-sgx", "dotnet", "publish/dotnetcoresample.dll"]
```

### Create `Makefile`

```
$ cat Makefile
ARCH_LIBDIR ?= /lib/$(shell $(CC) -dumpmachine)

ifeq ($(DEBUG),1)
GRAMINE_LOG_LEVEL = all
else
GRAMINE_LOG_LEVEL = error
endif

.PHONY: all
all: dotnet.manifest
ifeq ($(SGX),1)
all: dotnet.manifest.sgx dotnet.sig dotnet.token
endif

dotnet.manifest: dotnet.manifest.template
        gramine-manifest \
                -Dlog_level=$(GRAMINE_LOG_LEVEL) \
                -Darch_libdir=$(ARCH_LIBDIR) \
                -Dentrypoint=$(realpath $(shell sh -c "command -v dotnet")) \
                $< $@

# gramine-sgx-sign generates both a .sig file and a .manifest.sgx file. This is somewhat
# hard to express properly in Make. The simple solution would be to use
# "Rules with Grouped Targets" (`&:`), however make on Ubuntu <= 20.04 doesn't support it.
#
# Simply using a normal rule with "two targets" is equivalent to creating separate rules
# for each of the targets, and when using `make -j`, this might cause two instances
# of gramine-sgx-sign to get launched simultaneously, potentially breaking the build.
#
# As a workaround, we use a dummy intermediate target, and mark both files as depending on it, to
# get the dependency graph we want. We mark this dummy target as .INTERMEDIATE, which means
# that make will consider the source tree up-to-date even if the sgx_sign file doesn't exist,
# as long as the other dependencies check out. This is in contrast to .PHONY, which would
# be rebuilt on every invocation of make.
dotnet.sig dotnet.manifest.sgx: sgx_sign
        @:

.INTERMEDIATE: sgx_sign
sgx_sign: dotnet.manifest
        gramine-sgx-sign \
                --manifest $< \
                --output $<.sgx

dotnet.token: dotnet.sig
        gramine-sgx-get-token \
                --output $@ --sig $<

ifeq ($(SGX),)
GRAMINE = gramine-direct
else
GRAMINE = gramine-sgx
endif

.PHONY: check
check: all
        $(GRAMINE) dotnet publish/dotnetcoresample.dll

.PHONY: clean
clean:
        $(RM) *.token *.sig *.manifest.sgx *.manifest OUTPUT

.PHONY: distclean
distclean: clean
```

### `manifest` for dotnet

```
$ cat dotnet.manifest.template
loader.entrypoint = "file:{{ gramine.libos }}"
libos.entrypoint = "{{ entrypoint }}"

loader.log_level = "{{ log_level }}"

loader.pal_internal_mem_size = "128M"
loader.insecure__use_cmdline_argv = true

loader.env.LD_LIBRARY_PATH = "/lib:{{ arch_libdir }}:/usr/lib:/usr/{{ arch_libdir }}"
loader.env.DOTNET_GCHeapHardLimit = "0xC800000"    # 200 mebibytes (MiB)
loader.env.DOTNET_USE_POLLING_FILE_WATCHER = "1"
loader.env.DOTNET_gcServer = "1"
loader.env.DOTNET_RUNNING_IN_CONTAINER = "true"

sys.insecure__allow_eventfd = true

fs.mounts = [
  { uri = "file:{{ gramine.runtimedir() }}", path = "/lib" },
  { uri = "file:{{ arch_libdir }}", path = "{{ arch_libdir }}" },
  { uri = "file:/usr", path = "/usr" },
  { uri = "file:/etc", path = "/etc" },
  { type = "tmpfs", path = "/tmp" },
]

sgx.enclave_size = "4G"
sgx.thread_num = 128
sgx.debug = true

sgx.trusted_files = [
  "file:{{ gramine.libos }}",
  "file:{{ entrypoint }}",
  "file:{{ gramine.runtimedir() }}/",
  "file:{{ arch_libdir }}/",
  "file:/usr/{{ arch_libdir }}/",
  "file:/usr/share/dotnet/",
  "file:/usr/share/zoneinfo/",
  "file:/usr/share/zoneinfo-icu/",
  "file:publish/",
]

sgx.allowed_files = [
  # Name Service Switch (NSS) files. Glibc reads these files as part of name-
  # service information gathering. For more info, see 'man nsswitch.conf'.
  "file:/etc/nsswitch.conf",
  "file:/etc/hosts",
  "file:/etc/group",
  "file:/etc/passwd",

  "file:/etc/localtime",
  "file:/etc/"
]
```
### Clone the "hello world" sample code

```
$ git clone https://github.com/Azure-Samples/dotnetcore-docs-hello-world
```

### Build the docker image

```
$ docker build -t gramine-dotnetcore-docs-hello-world-linux .
```

### Run the image

```
$ docker run --device /dev/sgx_enclave --device /dev/sgx_provision -it gramine-dotnetcore-docs-hello-world-linux
```

### Expected results

The expectation of Gramine is to run ASP.NET app without any issue as the normal app. See the logs without Gramine:

```
$ dotnet publish/dotnetcoresample.dll
info: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[62]
      User profile is available. Using '/home/user/.aspnet/DataProtection-Keys' as key repository; keys will not be encrypted at rest.
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://localhost:5000
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:5001
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: /home/user/gramine-dotnetcore-docs-hello-world/
```

### Actual results

Exit with error. Partial logs:

```
...
[P1:T1:dotnet] debug: memory fault at 0x00000000 (IP = 0x158e7ac44)
[P1:T1:dotnet] debug: Created sigframe for sig: 11 at 0x158b29390 (handler: 0x15904cf60, restorer: 0x1594b2450)
warning: Unknown or illegal instruction executed at 0x159088f86
...
[P1:libos] debug: IPC worker: exiting worker thread
[P1:T1:dotnet] debug: process 1 exited with status 132
debug: PalProcessExit: Returning exit code 132
```

### Gramine commit hash

4f259b80ed4c40e7cccde70e822d378754dacb97

**Comment by kailun-qin** [comment]
> Have you tried gramine-direct?

`gramine-direct` can run it w/o a problem.

> I have a suspicion that ASP.NET tries to execute something like in or out instructions -- which are prohibited in Intel SGX.

Looks like the `iretq` in the dotnet runtime https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/coreclr/pal/src/arch/amd64/context2.S#L182 caused this.

**Comment by marcelamelara** [comment]
Hello! Wanted to ping this issue and ask if there are any updates. I've run into a similar error trying to run a dotnet application (GHA Runner) with `gramine-sgx`. The application also seems to loop infinitely in `gramine-direct`. 

Here's a log snippet from a run with `gramine-sgx`.

```
...
(libos_parser.c:1609:buf_write_all) [P1:T1:Runner.Listener] trace: ---- return from futex(...) = 0x0
(pal_exception.c:177:handle_ud) warning: Unknown or illegal instruction executed at 0x376762b6
(libos_signal.c:480:illegal_upcall) [P1:T1:Runner.Listener] debug: Illegal instruction during app execution at 0x376762b6; delivering to app
(libos_context.c:279:prepare_sigframe) [P1:T1:Runner.Listener] debug: Created sigframe for sig: 4 at 0x37ffe7d0 (handler: 0x37639c90, restorer: 0x37a9e580)
(libos_parser.c:1609:buf_write_all) [P1:T1:Runner.Listener] trace: ---- rt_sigprocmask(UNBLOCK, [SIGILL,], NULL, 0x8) = 0x0
(libos_parser.c:1609:buf_write_all) [P1:T1:Runner.Listener] trace: ---- rt_sigaction([SIGILL], 0x37ffe640, 0, 0x8) = 0x0
...
```

Unclear if it's due to `iretq`, so I'm happy to open a separate issue. Please let me know if that would be more helpful.

**Comment by monavij** [comment]
@marcelamelara  Maybe you should try PR #1166 to see if it's the same issue.

**Comment by marcelamelara** [comment]
Thanks @monavij for the suggestion. I was finally able to test this today, and #1166 did resolve these issues! I'm still having other problems with running my application in Gramine, but they seem unrelated to `iretq`.
